src/Ubidots.cpp

```diff
@@ -0,0 +1,160 @@
+  add(variable_label, value, context, dot_timestamp_seconds, NULL);
+}
+
+void Ubidots::add(const char* variable_label, float value, char* context, long unsigned dot_timestamp_seconds,
+                  unsigned int dot_timestamp_millis) {
+  _cloudProtocol->add(variable_label, value, context, dot_timestamp_seconds, dot_timestamp_millis);
+}
+
+/**
+ * Sends data to Ubidots
+ * @arg device_label [Mandatory] device label where the dot will be stored
+ * @arg device_name [optional] Name of the device to be created (supported only
+ * for TCP/UDP)
+ * @arg flags [Optional] Particle publish flags for webhooks
+ */
+
+bool Ubidots::send() { return _cloudProtocol->send(); }
+
+bool Ubidots::send(const char* device_label) { return _cloudProtocol->send(device_label); }
+
+bool Ubidots::send(const char* device_label, const char* device_name) {
+  return _cloudProtocol->send(device_label, device_name);
+}
+
+float Ubidots::get(const char* device_label, const char* variable_label) {
+  _cloudProtocol->get(device_label, variable_label);
+}
+
+void Ubidots::setDebug(bool debug) {
+  _debug = debug;
+  _cloudProtocol->setDebug(debug);
+}
+
+/*
+ * Adds to the context structure values to retrieve later it easily by the user
+ */
+
+void Ubidots::addContext(char* key_label, char* key_value) {
+  (_context + _current_context)->key_label = key_label;
+  (_context + _current_context)->key_value = key_value;
+  _current_context++;
+  if (_current_context >= MAX_VALUES) {
+    Serial.println(F("You are adding more than the maximum of consecutive key-values pairs"));
+    _current_context = MAX_VALUES;
+  }
+}
+
+/*
+ * Retrieves the actual stored context properly formatted
+ */
+
+void Ubidots::getContext(char* context_result) { getContext(context_result, _iotProtocol); }
+
+void Ubidots::getContext(char* context_result, IotProtocol iotProtocol) {
+  // TCP context type
+  if (iotProtocol == UBI_TCP || iotProtocol == UBI_UDP) {
+    sprintf(context_result, "");
+    for (uint8_t i = 0; i < _current_context;) {
+      sprintf(context_result, "%s%s=%s", context_result, (_context + i)->key_label, (_context + i)->key_value);
+      i++;
+      if (i < _current_context) {
+        sprintf(context_result, "%s$", context_result);
+      } else {
+        sprintf(context_result, "%s", context_result);
+        _current_context = 0;
+      }
+    }
+  }
+
+  // HTTP context type
+  if (iotProtocol == UBI_HTTP) {
+    sprintf(context_result, "");
+    for (uint8_t i = 0; i < _current_context;) {
+      sprintf(context_result, "%s\"%s\":\"%s\"", context_result, (_context + i)->key_label, (_context + i)->key_value);
+      i++;
+      if (i < _current_context) {
+        sprintf(context_result, "%s,", context_result);
+      } else {
+        sprintf(context_result, "%s", context_result);
+        _current_context = 0;
+      }
+    }
+  }
+}
+
+bool Ubidots::wifiConnect(const char* ssid, const char* password) {
+  return _cloudProtocol->wifiConnect(ssid, password);
```

Comment:
This method does not exist in the Abstract Class

Reply:
handled now by Ubidots